### PR TITLE
KAFKA-16921: Migrate test of connect module to Junit5 (Runtime direct)

### PR DESCRIPTION
**Pull request by gongxuanzhang**
This PR is part of a task submodule, with the purpose of migrating from JUnit 4 to JUnit 5.
see 
* https://github.com/apache/kafka/pull/16253
* https://github.com/apache/kafka/pull/16330
* https://github.com/apache/kafka/pull/16328

In this PR,AbstractWorkerSourceTaskTest  test case has error.maybe other modules are involved.

### CODE REVIEW

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java [hunk 1]
```diff
@@ -293,8 +287,10 @@ private void createWorkerTask(TargetState initialState, Converter keyConverter,
                 sourceConfig, Runnable::run, preProducerCheck, postProducerCheck, Collections::emptyList);
     }
 
-    @Test
-    public void testRemoveMetrics() {
+    @ParameterizedTest
+    @MethodSource("parameters")
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java]: Could you please use `@ValueSource(booleans = {true, false})` instead? that is more readable to me

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java [hunk 2]
```diff
@@ -310,8 +306,10 @@ private Set<MetricName> filterToTaskMetrics(Set<MetricName> metricNames) {
                 .collect(Collectors.toSet());
     }
 
-    @Test
-    public void testStartPaused() throws Exception {
+    @ParameterizedTest
+    @MethodSource("parameters")
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java]: ditto

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java [hunk 3]
```diff
@@ -333,8 +331,10 @@ public void testStartPaused() throws Exception {
         assertPollMetrics(0);
     }
 
-    @Test
-    public void testPause() throws Exception {
+    @ParameterizedTest
+    @MethodSource("parameters")
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java]: ditto

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java [hunk 4]
```diff
@@ -372,8 +372,10 @@ public void testPause() throws Exception {
         verifyCleanShutdown();
     }
 
-    @Test
-    public void testFailureInPreProducerCheck() throws Exception {
+    @ParameterizedTest
+    @MethodSource("parameters")
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java]: ditto

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java [hunk 5]
```diff
@@ -386,8 +388,10 @@ public void testFailureInPreProducerCheck() throws Exception {
         verifyShutdown(true, false);
     }
 
-    @Test
-    public void testFailureInProducerInitialization() throws Exception {
+    @ParameterizedTest
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java:386]: ditto

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java [hunk 6]
```diff
@@ -429,8 +446,10 @@ public void testShutdownFailure() {
         verifyShutdown(false, true);
     }
 
-    @Test
-    public void testTransitionStartedToStarted() {
+    @ParameterizedTest
+    @MethodSource("parameters")
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java]: Could you please use `@EnumSource(value = ConnectorType.class, names = {"SOURCE", "SINK"})` instead? it is more readable

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java [hunk 7]
```diff
@@ -355,8 +353,10 @@ public void testPollsInBackground() throws Exception {
         verifyClose();
     }
 
-    @Test
-    public void testFailureInPoll() throws Exception {
+    @ParameterizedTest
+    @MethodSource("parameters")
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for this patch. `MethodSource` is not readable, so we should take a look at both `ValueSource` and `EnumSource` first.

**Comment by gongxuanzhang** [comment]
@chia7712  I update it

**Comment by gongxuanzhang** [comment]
@chia7712  I update it

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ErrorHandlingTaskTest.java [hunk 8]
```diff
@@ -179,19 +176,15 @@ public class ErrorHandlingTaskTest {
 
     private ErrorHandlingMetrics errorHandlingMetrics;
 
-    private final boolean enableTopicCreation;
+    private boolean enableTopicCreation;
 
-    @Parameterized.Parameters
     public static Collection<Boolean> parameters() {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ErrorHandlingTaskTest.java]: this is unused now, right?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java [hunk 9]
```diff
@@ -174,23 +171,20 @@ public class ExactlyOnceWorkerSourceTaskTest {
...
+    private boolean enableTopicCreation;
 
     private boolean taskStarted;
     private Future<?> workerTaskFuture;
 
-    @ParameterizedTest.Parameters
-    public static Collection<Boolean> parameters() {
-        return Arrays.asList(false, true);
+    
+    public static Stream<Boolean> parameters() {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ExactlyOnceWorkerSourceTaskTest.java]: ditto

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java [hunk 10]
```diff
@@ -77,25 +75,22 @@ public class WorkerConnectorTest {
...
-    private final Connector connector;
-    private final CloseableOffsetStorageReader offsetStorageReader;
-    private final ConnectorOffsetBackingStore offsetStore;
+    private ConnectorType connectorType;
+    private Connector connector;
+    private CloseableOffsetStorageReader offsetStorageReader;
+    private ConnectorOffsetBackingStore offsetStore;
 
-    @Parameterized.Parameters
     public static Collection<ConnectorType> parameters() {
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR

**Comment by gongxuanzhang** [comment]
plz take a look @chia7712

**Comment by chia7712** [comment]
@gongxuanzhang please fix those failed tests

**Comment by gongxuanzhang** [comment]
I have fixed all the test cases that failed due to the changes.
@chia7712

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskThreadedTest.java [hunk 11]
```diff
@@ -643,7 +647,11 @@ public Object answer(InvocationOnMock invocation) {
                 }
             }
         }).when(sinkTask).preCommit(anyMap());
-
+        //  if the only command has error, consumer don't commit, so we don't need to mock commitAsync.
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskThreadedTest.java]: Could we have two methods? 

```java
    private void expectPreCommit(ExpectOffsetCommitCommand... commands) {
        doAnswer(new Answer<Object>() {
            int index = 0;

            @Override
            public Object answer(InvocationOnMock invocation) {
                ExpectOffsetCommitCommand commitCommand = commands[index++];
                // All assigned partitions will have offsets committed, but we've only processed messages/updated offsets for one
                final Map<TopicPartition, OffsetAndMetadata> offsetsToCommit = offsetsToCommitFn.apply(commitCommand.expectedMessages);

                if (commitCommand.error != null) {
                    throw commitCommand.error;
                } else {
                    return offsetsToCommit;
                }
            }
        }).when(sinkTask).preCommit(anyMap());
    }

    private void expectOffsetCommit(ExpectOffsetCommitCommand... commands) {
        expectPreCommit(commands);
        ...
   }
```

with that changes, each test case can call specific "expect"

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java [hunk 12]
```diff
@@ -190,19 +187,10 @@ public class WorkerSourceTaskTest {
             new SourceRecord(PARTITION, OFFSET, "topic", null, KEY_SCHEMA, KEY, RECORD_SCHEMA, RECORD)
     );
 
-    private final boolean enableTopicCreation;
-
-    @ParameterizedTest.Parameters
-    public static Collection<Boolean> parameters() {
-        return Arrays.asList(false, true);
-    }
+    private boolean enableTopicCreation;
```
Comment (chia7712) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSourceTaskTest.java]: we should remove this, and each test case should call `expectTopicCreation` when `enableTopicCreation=true`

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR

**Comment by gongxuanzhang** [comment]
I update it. plz take a look @chia7712

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**